npc/namespace.go

```diff
@@ -355,10 +494,61 @@ func isDefaultDeny(namespace *coreapi.Namespace) bool {
 		*(nnp.Ingress.Isolation) == DefaultDeny
 }
 
+func hasIP(pod *coreapi.Pod) bool {
+	// Ensure pod isn't dead, has an IP address and isn't sharing the host network namespace
+	return pod.Status.Phase != "Succeeded" && pod.Status.Phase != "Failed" &&
+		len(pod.Status.PodIP) > 0 && !pod.Spec.HostNetwork
+}
+
+func equals(a, b map[string]string) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for ak, av := range a {
+		if b[ak] != av {
+			return false
+		}
+	}
+	return true
+}
+
 func namespaceComment(namespace *ns) string {
 	return "namespace: " + namespace.name
 }
 
 func podComment(pod *coreapi.Pod) string {
 	return fmt.Sprintf("namespace: %s, pod: %s", pod.ObjectMeta.Namespace, pod.ObjectMeta.Name)
 }
+
+func (ns *ns) analyse(obj interface{}) (
+	uid types.UID,
+	rules map[string]*ruleSpec,
+	nsSelectors, podSelectors map[string]*selectorSpec,
+	err error) {
+
+	switch p := obj.(type) {
+	case *extnapi.NetworkPolicy:
+		uid = p.ObjectMeta.UID
+	case *networkingv1.NetworkPolicy:
+		uid = p.ObjectMeta.UID
+	default:
+		err = errInvalidNetworkPolicyObjType
+		return
+	}
+	ns.policies[uid] = obj
+
+	// Analyse policy, determine which rules and ipsets are required
+	if ns.legacy {
+		rules, nsSelectors, podSelectors, err = ns.analysePolicyLegacy(obj.(*extnapi.NetworkPolicy))
```

Comment:
Looks like this will crash if a v1 policy comes in in legacy mode, or vice-versa.  But maybe this cannot happen because of the way we subscribe?
Maybe this if-statement should be merged with the above switch, so it's clear that things can only happen one way?

Reply:
This cannot happen due to the way we subscribe, as you said. 

Maybe it makes sense to cast netpol objects directly w/o the type switch (inside `if ns.legacy { ... } else { ...}`). That would make the runtime panic if invalid object is given, but it's fine as we are no longer able to progress in such situation. WDYT?